cgaal-engine/src/parsing/parser.rs

```diff
@@ -87,8 +90,337 @@ impl<'a> Parser<'a> {
+                Some(tok) => {
+                    self.errors.log(
+                        tok.span,
+                        format!(
+                            "Unexpected '{}', expected '{}' or '{}'",
+                            tok.kind,
+                            TokenKind::Comma,
+                            self.recovery_tokens.last().unwrap()
+                        ),
+                    );
+                    return Err(RecoverMode);
+                }
+                None => {
+                    self.errors.log_msg(format!(
+                        "Unexpected EOF, expected '{}' or '{}'",
+                        TokenKind::Comma,
+                        self.recovery_tokens.last().unwrap()
+                    ));
+                    return Err(RecoverMode);
+                }
+            }
+        }
+        Ok(cases)
+    }
+
+    /// Parse a template declaration.
+    pub fn template_decl(&mut self) -> Result<Decl, RecoverMode> {
+        let start = self.token(TokenKind::KwTemplate)?;
+        let (end, (ident, decls)) = recover!(
+            self,
+            self.template_inner(),
+            TokenKind::KwEndTemplate,
+            (Ident::new_error(), Vec::new())
+        )?;
+        Ok(Decl::new(start + end, ident, DeclKind::Template(decls)))
+    }
+
+    /// Parse the inners a template declaration.
+    fn template_inner(&mut self) -> Result<(Ident, Vec<Decl>), RecoverMode> {
+        let ident = self.ident()?;
+        let decls = self.decls(true)?;
+        Ok((ident, decls))
+    }
+
+    /// Parse an action declaration.
+    pub fn action_decl(&mut self) -> Result<Decl, RecoverMode> {
+        let start = self.token(TokenKind::Lbracket)?;
+        let (_, ident) = recover!(self, self.ident(), TokenKind::Rbracket, Ident::new_error())?;
+        let cond = self.expr()?;
+        let span = start + cond.span;
+        let action = DeclKind::Action(Arc::new(cond));
+        Ok(Decl::new(span, ident, action))
+    }
+
+    /// Parse an expression.
+    pub fn expr(&mut self) -> Result<Expr, RecoverMode> {
+        // The sub-expressions of a ternary cannot be another ternary in order to avoid ambiguity.
+        // E.g. "a ? b : c ? d : f" is ambiguous and disallowed.
+        let cond = self.binary_expr(0)?;
+        match self.lexer.peek() {
+            Some(Token {
+                kind: TokenKind::Question,
+                ..
+            }) => {
+                self.lexer.next().unwrap();
+                let (_, then) = recover!(
+                    self,
+                    self.binary_expr(0),
+                    TokenKind::Colon,
+                    Expr::new_error()
+                )?;
+                let els = self.binary_expr(0)?;
+                let span = cond.span + els.span;
+                let kind = ExprKind::TernaryIf(cond.into(), then.into(), els.into());
+                Ok(Expr::new(span, kind))
+            }
+            _ => Ok(cond),
+        }
+    }
```

Comment:
What if we have parenthesis in the sub-expr of a ternary?
```
a ? (x ? y : z) : c
```
This should be allowed.

Reply:
Yes. This is allowed already. Ternary expressions consist of binary expressions which consist of terms. Parenthesis is a term that allows ternary expressions inside, making your snippet ok.